prov/tcp/src/xnet_rdm.c

```diff
@@ -695,13 +695,183 @@ static struct fi_ops_ep xnet_rdm_ep_ops = {
+	struct xnet_domain *subdomain = context;
+
+	ret = ofi_mr_map_insert(&subdomain->util_domain.mr_map, attr,
+				&attr->requested_key, attr->context,
+				((struct ofi_mr*)attr->context)->flags);
+	if (ret)
+		XNET_WARN_ERR(FI_LOG_MR, "ofi_mr_map_insert", ret);
+	else
+		ofi_atomic_inc32(&subdomain->util_domain.ref);
+}
+
+struct xnet_domain *find_subdomain(struct xnet_rdm *rdm)
+{
+	int i;
+	struct util_cntr *cntr;
+
+	if (!xnet_domain_multiplexed(&rdm->util_ep.rx_cq->domain->domain_fid)) {
+		return container_of(&rdm->util_ep.rx_cq->domain->domain_fid,
+				    struct xnet_domain, util_domain.domain_fid);
+	}
+
+	if (!xnet_domain_multiplexed(&rdm->util_ep.tx_cq->domain->domain_fid)) {
+		return container_of(&rdm->util_ep.tx_cq->domain->domain_fid,
+					 struct xnet_domain, util_domain.domain_fid);
+	}
+
+	for (i = 0; i < CNTR_CNT; i++) {
+		cntr = rdm->util_ep.cntrs[i];
+		if (!cntr)
+			continue;
+
+		if (!xnet_domain_multiplexed(&cntr->domain->domain_fid)) {
+			return container_of(&cntr->domain->domain_fid,
+					 struct xnet_domain, util_domain.domain_fid);
+		}
+	}
+
+	return NULL;
+}
+
+static void set_subdomain(struct xnet_rdm *rdm, struct xnet_domain *domain,
+			  struct xnet_domain *subdomain)
+{
+	int i;
+	struct util_cntr *cntr;
+
+	if (rdm->util_ep.rx_cq->domain == &domain->util_domain) {
+		ofi_atomic_dec32(&rdm->util_ep.rx_cq->domain->ref);
+		ofi_atomic_inc32(&subdomain->util_domain.ref);
+		rdm->util_ep.rx_cq->domain = &subdomain->util_domain;
+	}
+
+	if (rdm->util_ep.tx_cq->domain == &domain->util_domain) {
+		ofi_atomic_dec32(&rdm->util_ep.tx_cq->domain->ref);
+		ofi_atomic_inc32(&subdomain->util_domain.ref);
+		rdm->util_ep.tx_cq->domain = &subdomain->util_domain;
+	}
+
+	for (i = 0; i < CNTR_CNT; i++) {
+		cntr = rdm->util_ep.cntrs[i];
+		if (!cntr)
+			continue;
+
+		if (cntr->domain == &domain->util_domain) {
+			ofi_atomic_dec32(&cntr->domain->ref);
+			ofi_atomic_inc32(&subdomain->util_domain.ref);
+			cntr->domain = &subdomain->util_domain;
+		}
+	}
+}
+
+int xnet_rdm_resolve_domains(struct xnet_rdm *rdm)
+{
+	int ret;
+	struct fid_domain *subdomain_fid;
+	struct xnet_domain *subdomain;
+	struct xnet_domain *domain = container_of(rdm->util_ep.domain,
+							struct xnet_domain,
+							util_domain);
+	subdomain = find_subdomain(rdm);
+	if (!subdomain) {
+		ret = fi_domain(&domain->util_domain.fabric->fabric_fid,
+				domain->subdomain_info,
+				&subdomain_fid, NULL);
+		if (ret)
+			return ret;
+
+		subdomain = container_of(subdomain_fid, struct xnet_domain,
+					 util_domain.domain_fid);
+		ret = fid_list_insert2(&domain->subdomain_list,
+				       &domain->subdomain_list_lock,
+				       &subdomain_fid->fid);
+		if (ret)
+			fi_close(&subdomain_fid->fid);
+
+		set_subdomain(rdm, domain, subdomain);
+
+		ofi_rbmap_foreach(domain->util_domain.mr_map.rbtree,
+				  domain->util_domain.mr_map.rbtree->root,
+				  xnet_reg_subdomain_mr, subdomain);
+	}
+
+	ret = xnet_mplex_av_dup(&rdm->util_ep, domain, subdomain);
```

Comment:
If the subdomain already exists, hasn't the AV already been duplicated?

Reply:
yes I suppose this should move up inside the if block

Reply:
So, I guess it's not as simple as moving this up.  The av may truly be shared amongst eps, and in that case we likely don't want to duplicate but just update the references.  Instead, we likely want to check that a matching subav exists for the subdomain and only duplicate if it doesn't.

In attempting to resolve this, I discovered a case that must have escaped prior testing.  When the main av is populated prior to creating subavs, we iterate the av entry pool to insert into the new subav, however if there are regions in the main av that are not allocated, we will fail assertions in the underlying `ofi_bufpool_get_ibuf`.  I believe the options here are to relax the assertion, promote the `ofi_bufpool_hdr:allocated` flag from debug only, or to have some fairly significant code duplication.  I'm trying to find a better option before committing to any of those.

Reply:
Why would there be regions in the main AV that aren't allocated?  Isn't the AV created using AV_TABLE semantics, which would populate in order?

Reply:
we could overwrite the pool's chunk_cnt for the main av and set it to 1.  it would mean we have to call "ofi_bufpool_grow" more often, but no other code would need to change.

Reply:
I don't think we should interact directly with bufpool.  The AV APIs, even internal calls, should be usable, or updated to be usable.  We can use ofi_av_size() to get the size of the AV.  ofi_av_get_addr() should return the address for a given index (fi_addr_t), or NULL, I guess, if one does not exist.

Reply:
we are using the ofi_av calls.  the problem is that these calls (`ofi_av_get_addr`) call into `ofi_bufpool_get_ibuf` where we hit the assert if the debug `allocated` flag is false.  This `allocated` flag is the only indication that the buffer is _actually_ used. `ofi_av_size` returns the number of entries in the av, but doesn't tell you the number of entries that are actually used (marked `allocated`).

With all that said, it seems you may be leading me toward promoting that `allocated` flag so that we can rely on it to tell us the entries that are used rather than just for debug purposes.  this means that all providers that use the `util_av` will pick up that overhead (1 byte per entry?).